ir/src/integrity_stmts/mod.rs

```diff
@@ -123,3 +124,30 @@ impl IntegrityStmts {
         Ok(())
     }
 }
+
+/// A [ConstraintRoot] represents the entry node of a subgraph representing an integrity constraint
+/// within the [AlgebraicGraph]. It also contains the row offset for the constraint which is the
+/// maximum of all row offsets accessed by the constraint.
+#[derive(Debug, Clone)]
+pub struct ConstraintRoot {
+    index: NodeIndex,
+    offset: usize,
+}
+
+impl ConstraintRoot {
+    /// Creates a new [ConstraintRoot] with the specified entry index and row offset.
+    pub fn new(index: NodeIndex, offset: usize) -> Self {
+        Self { index, offset }
+    }
+
+    /// Returns the index of the entry node of the subgraph representing the constraint.
+    pub fn node_index(&self) -> &NodeIndex {
+        &self.index
+    }
+
+    /// Returns the row offset for the constraint which is the maximum of all the row offsets
+    /// accessed by the constraint.
+    pub fn offset(&self) -> usize {
+        self.offset
+    }
```

Comment:
I think this row offset terminology may not be entirely clear from these comments, since I don't think we explain it fully anywhere in the code. Maybe we should include an example like "For example, if a constraint only accesses the trace in the current row then the row offset will be 0, but if it accesses the trace in both the current and the next rows then the row offset will be 1."

Reply:
Thanks. Done.